File: csharp/failover.cs
using System;
using System.Text;
using System.Security.Cryptography;
using System.Threading.Tasks;
using System.Collections.Generic;

// imports for Azure IoT
using Microsoft.Azure.Devices.Client;
using Microsoft.Azure.Devices.Provisioning.Client;
using Microsoft.Azure.Devices.Provisioning.Client.Transport;
using Microsoft.Azure.Devices.Shared;

namespace failover
{
    class Program
    {

        // device settings - FILL IN YOUR VALUES HERE
        private static string scopeId = "<Put your scope id here from IoT Central Administration -> Device connection>";
        private static string groupSymmetricKey = "<Put your group SAS primary key here from IoT Central Administration -> Device Connection -> SAS-IoT-Devices>";

        // optional device settings - CHANGE IF DESIRED/NECESSARY
        private static string deviceId = "failover3_cs";
        private static string globalDeviceEndpoint = "global.azure-devices-provisioning.net";
        private static string modelId = "dtmi:Sample:Failover;1";
 
        // test setting flags
        private static bool telemetrySendOn = true;
        private static bool reportedPropertySendOn = true;
        private static bool desiredPropertyReceiveOn = true;
        private static bool directMethodReceiveOn = true;
        private static bool c2dCommandReceiveOn = true;

        // general purpose variables
        private static DeviceClient iotClient;
        private static bool connected = false;
        private static bool terminate = false;


        // calculate the device key using the symetric group key
        private static string ComputeDerivedSymmetricKey(string enrollmentKey, string deviceId)
        {
            if (string.IsNullOrWhiteSpace(enrollmentKey))
            {
                return enrollmentKey;
            }

            using var hmac = new HMACSHA256(Convert.FromBase64String(enrollmentKey));
            return Convert.ToBase64String(hmac.ComputeHash(Encoding.UTF8.GetBytes(deviceId)));
        }


        // monitor for the ESC key to be pressed for exiting the program
        public static void MonitorKeypress()
        {
            ConsoleKeyInfo cki = new ConsoleKeyInfo();
            do 
            {
                cki = Console.ReadKey(true);
            } while (cki.Key != ConsoleKey.Escape);   
            terminate = true;   
        }

        
        // handler for disconnects, reconnect via DPS
        public static async void ConnectionStatusChanges(ConnectionStatus status, ConnectionStatusChangeReason reason) {
            Console.WriteLine($"Connection Status: {status}");
            if (status == ConnectionStatus.Connected) {
                connected = true;
            } else if (status == ConnectionStatus.Disconnected) {
                // we are disconnected, need to reconnect via DPS
                connected = false;
                if (iotClient != null) {
                    await iotClient.CloseAsync();
                }
                await Connect();
            }
        }


        // handles desired properties from IoT Central (or hub)
        private static async Task OnDesiredPropertyChangedAsync(TwinCollection desiredProperties, object userContext)
        {
            if (desiredProperties.Contains("fanSpeed")) {
                Console.WriteLine($"Fan speed change value from IoT Central: {desiredProperties["fanSpeed"]}");

                // acknowledge the desired property back to IoT Central
                var reportedProperties = new TwinCollection($"{{\"fanSpeed\":{{\"value\": {desiredProperties["fanSpeed"]}, \"ac\":200, \"ad\":\"completed\", \"av\":{desiredProperties["$version"]}}}}}");
                await iotClient.UpdateReportedPropertiesAsync(reportedProperties);                
            } else {
                Console.WriteLine($"Unknown desired property: {desiredProperties.ToJson()}");
            }
        }


        // handles direct method 'echo' from IoT Central (or hub)
        private static async Task<MethodResponse> OnEchoMessageReceivedAsync(MethodRequest methodRequest, object userContext) {
            Console.WriteLine($"Echo command received with parameter {Encoding.UTF8.GetString(methodRequest.Data)}");
            
            // acknowledge the direct method
            return new MethodResponse(methodRequest.Data, 200); 
        }

        // handles the Cloud to Device (C2D) message setAlarm
        private static async Task OnC2dMessageReceivedAsync(Message receivedMessage, object userContext) {
            Console.WriteLine($"Received C2D message {receivedMessage.Properties["method-name"]}");
            if (receivedMessage.Properties["method-name"] == "setAlarm") {
                Console.WriteLine($"Setting an alarm for {Encoding.ASCII.GetString(receivedMessage.GetBytes())}");
            }
            receivedMessage.Dispose();
        }


        // connect to IoT Central/Hub via Device Provisioning Servicee (DPS)
        public static async Task Connect() {
            // calc device symmetric key from group symmetric key
            string deviceKey = ComputeDerivedSymmetricKey(groupSymmetricKey, deviceId);

            Console.WriteLine($"Initializing the device provisioning client...");
            // DPS provision with device symmetric key
            using var security = new SecurityProviderSymmetricKey(
                deviceId,
                deviceKey,
                null);

            using var transportHandler = new ProvisioningTransportHandlerMqtt(TransportFallbackType.WebSocketOnly);


            ProvisioningDeviceClient provClient = ProvisioningDeviceClient.Create(
                globalDeviceEndpoint,
                scopeId,
                security,
                transportHandler);
            
            Console.WriteLine($"Initialized for registration Id {security.GetRegistrationID()}.");

            // set the model to register against
            var modelPayload = new ProvisioningRegistrationAdditionalData
            {
                JsonData = $"{{ \"modelId\": \"{modelId}\" }}",
            };

            Console.WriteLine("Registering with the device provisioning service...");
            // register the device and get the hub host name
            DeviceRegistrationResult result = await provClient.RegisterAsync(modelPayload);

            Console.WriteLine($"Registration status: {result.Status}.");
            if (result.Status != ProvisioningRegistrationStatusType.Assigned)
            {
                Console.WriteLine($"Registration status did not assign a hub, so exiting this sample.");
                return;
            }

            Console.WriteLine($"Device {result.DeviceId} registered to {result.AssignedHub}.");

            Console.WriteLine("Creating symmetric key authentication for IoT Hub...");
            IAuthenticationMethod auth = new DeviceAuthenticationWithRegistrySymmetricKey(
                result.DeviceId,
                security.GetPrimaryKey());

            Console.WriteLine($"Testing the provisioned device with IoT Hub...");
            iotClient = DeviceClient.Create(result.AssignedHub, auth, TransportType.Mqtt);
            iotClient.SetConnectionStatusChangesHandler(ConnectionStatusChanges);

            // cannot use the default retry logic built into the SDK as it will not fallback to DPS
            iotClient.SetRetryPolicy(new NoRetry());
            
            // handler for desired twin properties
            if (desiredPropertyReceiveOn) {
                await iotClient.SetDesiredPropertyUpdateCallbackAsync(OnDesiredPropertyChangedAsync, null);
            }

            // handlers for the direct method
            if (directMethodReceiveOn) {
                await iotClient.SetMethodHandlerAsync("echo", OnEchoMessageReceivedAsync, null);
            }

            // handler for C2D messages
            if (c2dCommandReceiveOn) {
                await iotClient.SetReceiveMessageHandlerAsync(OnC2dMessageReceivedAsync, null);
            }
        }


        // sends telemetry on a set frequency
        public static async Task SendTelemetry(int sendFrequencyInSeconds) {
            var rand = new Random();
            while (!terminate) {
                if (connected) {
                    var telemetryPayload = $"{{\"temp\": {Math.Round(rand.NextDouble() * 100, 2)}, \"humidity\": {Math.Round(rand.NextDouble() * 100, 2)}}}";
                    Console.WriteLine($"Sending telemetry: {telemetryPayload}");
                    using var message = new Message(Encoding.UTF8.GetBytes(telemetryPayload));
                    message.ContentType = "application/json";
                    message.ContentEncoding = Encoding.UTF8.ToString();
                    await iotClient.SendEventAsync(message);
                }
                System.Threading.Thread.Sleep(sendFrequencyInSeconds * 1000);
            }
        }


        // sends reported properties on a set frequency
        public static async Task SendReportedProperty(int sendFrequencyInSeconds) {
            var rand = new Random();
            while (!terminate) {
                if (connected) {
                    var reportedProperties = new TwinCollection
                    {
                        ["battery"] = Math.Round(rand.NextDouble() * 100, 2)
                    };
                    Console.WriteLine($"Sending twin reported property: {reportedProperties.ToJson()}");

                    await iotClient.UpdateReportedPropertiesAsync(reportedProperties);
                }
                System.Threading.Thread.Sleep(sendFrequencyInSeconds * 1000);                
            }
        }


        // Entry point: Connect the device and start processing telemetry, properties and commands
        public static async Task<int> Main(string[] args)
        {
            // connect to IoT Hub via DPS
            await Connect();

            // keyboard monitor
            Console.WriteLine("\n\n*** Press the ESC key to exit from this program and shutdown things cleanly ***\n\n");
            Task consoleKeyTask = Task.Run(() => { MonitorKeypress(); });

            // start the telemetry and reported property send tasks
            List<Task> tasks = new List<Task>();
            if (telemetrySendOn) {
                tasks.Add(SendTelemetry(5));
            }
            if (reportedPropertySendOn) {
                tasks.Add(SendReportedProperty(10));
            }

            // await the termination of the telemetry and reported property send tasks
            await Task.WhenAll(tasks);

            // clean up
            Console.WriteLine("Cleaning up and exiting");
            await iotClient.CloseAsync();

            return 1;
        }
    }
}

File: Node/failover.js
'use strict';

// Azure IoT Device SDK
const Protocol  = require('azure-iot-device-mqtt').Mqtt;
const Client = require('azure-iot-device').Client;
const Message = require('azure-iot-device').Message;
const  ExponentialBackOffWithJitter = require('azure-iot-common').ExponentialBackOffWithJitter;

// Azure IoT DPS SDK
const ProvisioningTransport = require('azure-iot-provisioning-device-mqtt').Mqtt;
const SymmetricKeySecurityClient = require('azure-iot-security-symmetric-key').SymmetricKeySecurityClient;
const ProvisioningDeviceClient = require('azure-iot-provisioning-device').ProvisioningDeviceClient;

// Crypto SDK needed for computing device keys
const crypto = require('crypto');

// device settings - FILL IN YOUR VALUES HERE
const scopeId = "<Put your scope id here from IoT Central Administration -> Device connection>"
const groupSymmetricKey = "<Put your group SAS primary key here from IoT Central Administration -> Device Connection -> SAS-IoT-Devices>"

// optional device settings - CHANGE IF DESIRED/NECESSARY
const provisioningHost = "global.azure-devices-provisioning.net"
const deviceId = "failover_js"
const modelId = "dtmi:Sample:Failover;1"  // This model is available in the root of the Github repo (Failover.json) and can be imported into your Azure IoT central application

// test setting flags
const telemetrySendOn = true
const reportedPropertySendOn = true
const desiredPropertyReceiveOn = true
const directMethodReceiveOn = true
const c2dCommandReceiveOn = true

// general purpose variables
let client = null;
let deviceTwin = null;
let connected = false;

// calculate the device key using the symetric group key
function computeDerivedSymmetricKey(masterKey, deviceId) {
    return crypto.createHmac('SHA256', Buffer.from(masterKey, 'base64'))
      .update(deviceId, 'utf8')
      .digest('base64');
}

// Azure IoT Central custom retry policy derived from ExponentialBackOffWithJitter
class MultiHubRetryPolicy extends ExponentialBackOffWithJitter {
    constructor(...args) {
        super(...args);
    }

    shouldRetry(err) { 
        if (err.message === 'Connection refused: Server unavailable')
           return false; // if hub not available stop retry and fall back to DPS
        else
            return super.shouldRetry(err);
    }

    nextRetryTimeout(retryCount, throttled) {
        return super.nextRetryTimeout(retryCount, throttled);
    }
}

// handler for C2D message
function messageHandler(msg) {
    let methodName = msg.properties.propertyList.find(o => o.key === 'method-name');
    // is this the setAlarm C2D message
    if (methodName != null && methodName.value === 'setAlarm') {
        console.log(`C2D method: ${methodName.value}(${msg.data.toString('utf-8')})`)
    }
}

// connect to IoT Central/Hub via Device Provisioning Servicee (DPS)
async function connect() {
    return new Promise((myResolve, myReject) => {
        // calc device symmetric key from group symmetric key
        const deviceSymmetricKey = computeDerivedSymmetricKey(groupSymmetricKey, deviceId);
        
        // DPS provision with device symmetric key
        const provisioningSecurityClient = new SymmetricKeySecurityClient(deviceId, deviceSymmetricKey);
        const provisioningClient = ProvisioningDeviceClient.create(provisioningHost, scopeId, new ProvisioningTransport(), provisioningSecurityClient);

        // set the model to register against
        provisioningClient.setProvisioningPayload(`{"iotcModelId":"${modelId}"}`)

        // register the device and get the hub host name
        provisioningClient.register(function(err, result) {
            if (err) {
                console.log("error registering device: " + err);
                throw new Error(`Registration error! Error: ${err.Message}`);
            } else {
                console.log('registration succeeded');
                console.log(`assigned hub: ${result.assignedHub}`);
                console.log(`deviceId: ${result.deviceId}`);
                const connectionString = `HostName=${result.assignedHub};DeviceId=${result.deviceId};SharedAccessKey=${deviceSymmetricKey}`;

                // create client from connection string
                client = Client.fromConnectionString(connectionString, Protocol);
                
                // cannot use the default retry logic built into the SDK as it will not fallback to DPS
                client.setRetryPolicy(new MultiHubRetryPolicy());

                // monitor for connects, disconnects, errors, and c2d messages
                client.on('connect', connectHandler);
                client.on('disconnect', disconnectHandler);
                client.on('error', errorHandler);
                if (c2dCommandReceiveOn)
                    client.on('message', messageHandler);

                // connect to IoT Hub
                client.open((err) => {
                    if (err) {
                        console.error(`Could not connect: ${err.message}`);
                        throw new Error(`Hub connect error! Error: ${err.Message}`);
                    } else {
                        // obtain twin object
                        client.getTwin(function(err, twin) {
                            if (err) {
                                throw new Error(`Obtaining twin error! Error: ${err.Message}`);
                            } else {
                                deviceTwin = twin
                                if (desiredPropertyReceiveOn) {
                                    deviceTwin.on('properties.desired', desiredPropertyHandler);
                                }
                                myResolve();
                            }
                        });
                    }
                });
            }
        });
    });
}

// handlef for connection event
function connectHandler() {
    console.log('Connected to IoT Central');  
    connected = true;
}

// handler for disconnects, reconnect via DPS
async function disconnectHandler() {
    if (connected) {
        connected = false;
        console.log('Disconnected from IoT Central');         
        client.close()
        await connect();
    }
}

// handler for errors
function errorHandler(err) {
    console.log(`Error caught in error handler: ${err}`);
}

// sends telemetry on a set frequency
async function sendTelemetry() {
    if (connected) {
        const telemetry = {"temp": (20 + (Math.random() * 100)).toFixed(2), "humidity": (Math.random() * 100).toFixed(2)}
        const message = new Message(JSON.stringify(telemetry));
        client.sendEvent(message, (err, res) => {
            if (err) {
                console.log(`Error: ${err.toString()}`);
            } else {
                console.log(`Completed telemetry send ${JSON.stringify(telemetry)}`);
            }
        });
    }
}

// sends reported properties on a set frequency
async function sendReportedProperty() {
    if (connected) {
        const reportedPropertyPatch = {"battery": (Math.random() * 100).toFixed(2)};
        deviceTwin.properties.reported.update(reportedPropertyPatch, (err) => {
            if (err) {
                console.log(`Error: ${err.toString()}`);
            } else {
                console.log(`Completed reported property send ${JSON.stringify(reportedPropertyPatch)}`);
            }
        });
    }
}

// handles desired properties from IoT Central (or hub)
function desiredPropertyHandler(patch) {
    if (Object.keys(patch).length > 1) {
        console.log(`Desired property received, the data in the desired properties patch is: ${JSON.stringify(patch)}`);

        // acknowledge the desired property back to IoT Central
        let key = Object.keys(patch)[0];
        if (key == "$version") {
            key = Object.keys(patch)[1];
        }
        let reported_payload = {};
        reported_payload[key] = {"value": patch[key], "ac":200, "ad":"completed", "av":patch['$version']};
        deviceTwin.properties.reported.update(reported_payload, (err) => {
            if (err) {
                console.log(`Error sending reported property ${err}`);
            } else {
                console.log(`Completed desired property acknowledgment ${JSON.stringify(reported_payload)}`);
            }
        });
    }
}

// handles direct method 'echo' from IoT Central (or hub)
function echoCommandHandler(request, response) {
    console.log(`Executing direct method request: ${request.methodName}(${request.payload})`);
    // echos back the request payload
    response.send(200, request.payload, function(err) {
        if (err) {
            console.log('Error');
        }
    });
}

// handles the Cloud to Device (C2D) message setAlarm
function setAlarmCommandHandler(request, response) {
    console.log(`Executing C2D message request: ${request.methodName}(${request.payload})`)
    response.send(200, function(err) {
        if (err) {
            console.log('Error');
        }
    });
}


// Connect the device and start processing telemetry, properties and commands
(async () => {
    try {
        console.log('Press Ctrl-C to exit from this when running in the console');

        // connect to IoT Central/Hub via Device Provisioning Service (DPS)
        await connect();

        // handlers for the direct method
        if (directMethodReceiveOn) {
            client.onDeviceMethod('echo', echoCommandHandler);
        }

        // start the interval timers to send telemetry and reported properties
        let sendTelemetryLoop = null;
        let sendReportedPropertiesLoop = null;
        if (telemetrySendOn) {
            const sendTelemetryLoop = setInterval(sendTelemetry, 5000); // send telemetry every 5 seconds
        }
        if (reportedPropertySendOn) {
            const sendReportedPropertiesLoop = setInterval(sendReportedProperty, 15000);  // send reported property every 15 seconds
        }

        // exit handler for cleanup of resources
        function exitHandler(options, exitCode) {
            if (options.cleanup) {
                console.log('Cleaning up and exiting');
                if (sendTelemetryLoop !== null) {
                    clearInterval(sendTelemetryLoop);
                }
                if (sendReportedPropertiesLoop !== null) {
                    clearInterval(sendReportedPropertiesLoop);
                }
                client.close();
            }
            if (options.exit) process.exit();
        }

        // try and cleanly exit when ctrl-c is pressed
        process.on('exit', exitHandler.bind(null,{cleanup:true}));
        process.on('SIGINT', exitHandler.bind(null, {exit:true}));
        process.on('SIGUSR1', exitHandler.bind(null, {exit:true}));
        process.on('SIGUSR2', exitHandler.bind(null, {exit:true}));

    } catch(e) {
        console.log(`Error: ${e}`);
    }
})();
